Add <input type=checkbox switch> pointer tracking

annevk opened the pull request:
#### 870a0515b649fa5462225fbf33a150e41f1827b5
<pre>
Add &lt;input type=checkbox switch&gt; pointer tracking
<a href="https://bugs.webkit.org/show_bug.cgi?id=265658">https://bugs.webkit.org/show_bug.cgi?id=265658</a>

Reviewed by Aditya Keerthi.

This change enables the switch&apos;s thumb to be &quot;dragged&quot; left and right.

This requires the introduction of a visual on/off state as actual state
changes, i.e., to checkedness, are only made after the tracking stops.
As such we simplify RenderTheme::isChecked() back to how it was before
we started adding &lt;input type=checkbox switch&gt; as switches don&apos;t need
it.

Despite what 271327@main claimed this change makes CheckboxInputType
solely responsible for performSwitchCheckedChangeAnimation(). This
works by passing along the event&apos;s isTrusted() bit to
willUpdateCheckedness() and instead stopping the animation from there.

A future change might have to abstract some of the logic in
CheckboxInputType::handleMouseMoveEvent() somewhat as more platforms
become supported.

A difference with macOS switches is that these remain active/pressed
and responsive to mouseup events while tracking is ongoing. This
difference with is also there for checkboxes and radio buttons.

* LayoutTests/fast/forms/switch/pointer-tracking-disabled-expected.html: Added.
* LayoutTests/fast/forms/switch/pointer-tracking-disabled.html: Added.
* LayoutTests/fast/forms/switch/pointer-tracking-expected-mismatch.html: Added.
* LayoutTests/fast/forms/switch/pointer-tracking-there-and-back-again-expected.html: Added.
* LayoutTests/fast/forms/switch/pointer-tracking-there-and-back-again-rtl-expected.html: Added.
* LayoutTests/fast/forms/switch/pointer-tracking-there-and-back-again-rtl.html: Added.
* LayoutTests/fast/forms/switch/pointer-tracking-there-and-back-again.html: Added.
* LayoutTests/fast/forms/switch/pointer-tracking.html: Added.
* Source/WebCore/html/CheckboxInputType.cpp:
(WebCore::CheckboxInputType::handleMouseDownEvent):
(WebCore::CheckboxInputType::handleMouseMoveEvent):
(WebCore::CheckboxInputType::willDispatchClick):
(WebCore::CheckboxInputType::startSwitchPointerTracking):
(WebCore::CheckboxInputType::stopSwitchPointerTracking):
(WebCore::CheckboxInputType::isSwitchPointerTracking const):
(WebCore::CheckboxInputType::disabledStateChanged):
(WebCore::CheckboxInputType::willUpdateCheckedness):
(WebCore::CheckboxInputType::performSwitchCheckedChangeAnimation):
(WebCore::CheckboxInputType::isSwitchVisuallyOn const):
* Source/WebCore/html/CheckboxInputType.h:
* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::HTMLInputElement::setChecked):
(WebCore::HTMLInputElement::defaultEventHandler):
(WebCore::HTMLInputElement::switchCheckedChangeAnimationProgress const):
(WebCore::HTMLInputElement::isSwitchVisuallyOn const):
* Source/WebCore/html/HTMLInputElement.h:
* Source/WebCore/html/InputType.cpp:
(WebCore::InputType::handleClickEvent): Deleted.
(WebCore::InputType::handleMouseDownEvent): Deleted.
(WebCore::InputType::handleDOMActivateEvent): Deleted.
(WebCore::InputType::willDispatchClick): Deleted.
(WebCore::InputType::didDispatchClick): Deleted.
* Source/WebCore/html/InputType.h:
(WebCore::InputType::handleClickEvent):
(WebCore::InputType::handleMouseDownEvent):
(WebCore::InputType::handleMouseMoveEvent):
(WebCore::InputType::willDispatchClick):
(WebCore::InputType::didDispatchClick):
(WebCore::InputType::handleDOMActivateEvent):
(WebCore::InputType::willUpdateCheckedness):
* Source/WebCore/html/RadioInputType.cpp:
(WebCore::RadioInputType::willUpdateCheckedness):
* Source/WebCore/html/RadioInputType.h:
* Source/WebCore/html/RangeInputType.cpp:
(WebCore::RangeInputType::handleMouseDownEvent):
* Source/WebCore/platform/graphics/controls/SwitchThumbPart.h:
* Source/WebCore/platform/graphics/controls/SwitchTrackPart.h:
* Source/WebCore/platform/graphics/mac/controls/SwitchThumbMac.mm:
(WebCore::SwitchThumbMac::draw):
* Source/WebCore/platform/graphics/mac/controls/SwitchTrackMac.mm:
(WebCore::SwitchTrackMac::draw):
* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::updateSwitchThumbPartForRenderer):
(WebCore::updateSwitchTrackPartForRenderer):
(WebCore::RenderTheme::isChecked const):
* Source/WebCore/rendering/RenderTheme.h:
(WebCore::RenderTheme::switchPointerTrackingMagnitudeProportion const):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: <a href="https://commits.webkit.org/271542@main">https://commits.webkit.org/271542@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3a7701869a18be8f8501336b51a57f7c6af8ed24

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/28721 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/7364 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/30092 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/31340 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/26205 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/29291 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/9487 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/4724 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/26274 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/28991 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/6092 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/24686 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/5291 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/5446 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/25690 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/32678 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/26290 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/26133 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/31683 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/5411 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/3583 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/29459 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/7029 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/25484 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6871 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/5879 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/5929 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

annevk commented on Source/WebCore/platform/graphics/mac/controls/SwitchThumbMac.mm:
```diff
@@ -58,7 +58,7 @@
 {
     LocalDefaultSystemAppearance localAppearance(style.states.contains(ControlStyle::State::DarkAppearance), style.accentColor);
 
-    bool isOn = style.states.contains(ControlStyle::State::Checked);
```
I just realized this means we can simplify RenderTheme::isChecked. (Alternatively we could make that take on this logic, but this seems cleaner? Not sure.)

pxlcoder commented on Source/WebCore/html/HTMLInputElement.cpp:
```diff
@@ -2312,4 +2310,10 @@ float HTMLInputElement::switchCheckedChangeAnimationProgress() const
+bool HTMLInputElement::isSwitchVisuallyOn() const
+{
+    ASSERT(isSwitch());
+    return downcast<CheckboxInputType>(*m_inputType).isSwitchVisuallyOn();
```
`checkedDowncast`

pxlcoder commented on Source/WebCore/html/CheckboxInputType.h:
```diff
@@ -57,13 +57,23 @@ class CheckboxInputType final : public BaseCheckableInputType {
     void stopSwitchCheckedChangeAnimation();
     void switchCheckedChangeAnimationTimerFired();
 
+    int m_switchPointerTrackingXPositionStart { -1 };
```
We generally prefer to use `std::optional`, rather than sentinel values.

pxlcoder commented on Source/WebCore/html/CheckboxInputType.h:
```diff
@@ -57,13 +57,23 @@ class CheckboxInputType final : public BaseCheckableInputType {
 
+    int m_switchPointerTrackingXPositionStart { -1 };
+    bool m_hasSwitchVisuallyOnChanged { false };
+    bool m_isSwitchVisuallyOn { false };
```
In the future, we should probably consider storing switch-specific data as a sidecar object, since most checkboxes don't need this information.

pxlcoder commented on Source/WebCore/html/CheckboxInputType.cpp:
```diff
@@ -87,6 +90,56 @@ void CheckboxInputType::handleKeyupEvent(KeyboardEvent& event)
+    auto changePosition = switchTrackWidth / 2;
+    if (!m_hasSwitchVisuallyOnChanged) {
+        auto switchTrackNoThumbWidth = switchTrackWidth - switchThumbLength;
+        auto changeOffset = switchTrackWidth * 0.4f;
```
Can you explain the `0.4`?

pxlcoder commented on LayoutTests/fast/forms/switch/pointer-tracking.html:
```diff
@@ -0,0 +1,14 @@
+    window.eventSender.mouseUp();
+}
+function end() {
+    setTimeout(() => document.documentElement.removeAttribute("class"), 50);
```
Is the `setTimeout` really necessary?

annevk replied:
It was for mac-wk1 at least.

pxlcoder commented on LayoutTests/fast/forms/switch/pointer-tracking-there-and-back-again.html:
```diff
@@ -0,0 +1,16 @@
+    window.eventSender.mouseUp();
+}
+function end() {
+    setTimeout(() => document.documentElement.removeAttribute("class"), 50);
```
Is the `setTimeout` really necessary?

annevk replied:
Yeah same as above.

pxlcoder commented on Source/WebCore/html/CheckboxInputType.h:
```diff
@@ -57,13 +57,25 @@ class CheckboxInputType final : public BaseCheckableInputType {
     void switchCheckedChangeAnimationTimerFired();
 
+    // FIXME: consider moving all switch-related state (and methods?) to their own object so
+    // CheckboxInputType can stay somewhat small
```
```suggestion
    // FIXME: Consider moving all switch-related state (and methods?) to their own object so
    // CheckboxInputType can stay somewhat small.
```

annevk committed:
Add <input type=checkbox switch> pointer tracking
https://bugs.webkit.org/show_bug.cgi?id=265658

Reviewed by Aditya Keerthi.

This change enables the switch's thumb to be "dragged" left and right.

This requires the introduction of a visual on/off state as actual state
changes, i.e., to checkedness, are only made after the tracking stops.
As such we simplify RenderTheme::isChecked() back to how it was before
we started adding <input type=checkbox switch> as switches don't need
it.

Despite what 271327@main claimed this change makes CheckboxInputType
solely responsible for performSwitchCheckedChangeAnimation(). This
works by passing along the event's isTrusted() bit to
willUpdateCheckedness() and instead stopping the animation from there.

A future change might have to abstract some of the logic in
CheckboxInputType::handleMouseMoveEvent() somewhat as more platforms
become supported.

A difference with macOS switches is that these remain active/pressed
and responsive to mouseup events while tracking is ongoing. This
difference with is also there for checkboxes and radio buttons.

* LayoutTests/fast/forms/switch/pointer-tracking-disabled-expected.html: Added.
* LayoutTests/fast/forms/switch/pointer-tracking-disabled.html: Added.
* LayoutTests/fast/forms/switch/pointer-tracking-expected-mismatch.html: Added.
* LayoutTests/fast/forms/switch/pointer-tracking-there-and-back-again-expected.html: Added.
* LayoutTests/fast/forms/switch/pointer-tracking-there-and-back-again-rtl-expected.html: Added.
* LayoutTests/fast/forms/switch/pointer-tracking-there-and-back-again-rtl.html: Added.
* LayoutTests/fast/forms/switch/pointer-tracking-there-and-back-again.html: Added.
* LayoutTests/fast/forms/switch/pointer-tracking.html: Added.
* Source/WebCore/html/CheckboxInputType.cpp:
(WebCore::CheckboxInputType::handleMouseDownEvent):
(WebCore::CheckboxInputType::handleMouseMoveEvent):
(WebCore::CheckboxInputType::willDispatchClick):
(WebCore::CheckboxInputType::startSwitchPointerTracking):
(WebCore::CheckboxInputType::stopSwitchPointerTracking):
(WebCore::CheckboxInputType::isSwitchPointerTracking const):
(WebCore::CheckboxInputType::disabledStateChanged):
(WebCore::CheckboxInputType::willUpdateCheckedness):
(WebCore::CheckboxInputType::performSwitchCheckedChangeAnimation):
(WebCore::CheckboxInputType::isSwitchVisuallyOn const):
* Source/WebCore/html/CheckboxInputType.h:
* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::HTMLInputElement::setChecked):
(WebCore::HTMLInputElement::defaultEventHandler):
(WebCore::HTMLInputElement::switchCheckedChangeAnimationProgress const):
(WebCore::HTMLInputElement::isSwitchVisuallyOn const):
* Source/WebCore/html/HTMLInputElement.h:
* Source/WebCore/html/InputType.cpp:
(WebCore::InputType::handleClickEvent): Deleted.
(WebCore::InputType::handleMouseDownEvent): Deleted.
(WebCore::InputType::handleDOMActivateEvent): Deleted.
(WebCore::InputType::willDispatchClick): Deleted.
(WebCore::InputType::didDispatchClick): Deleted.
* Source/WebCore/html/InputType.h:
(WebCore::InputType::handleClickEvent):
(WebCore::InputType::handleMouseDownEvent):
(WebCore::InputType::handleMouseMoveEvent):
(WebCore::InputType::willDispatchClick):
(WebCore::InputType::didDispatchClick):
(WebCore::InputType::handleDOMActivateEvent):
(WebCore::InputType::willUpdateCheckedness):
* Source/WebCore/html/RadioInputType.cpp:
(WebCore::RadioInputType::willUpdateCheckedness):
* Source/WebCore/html/RadioInputType.h:
* Source/WebCore/html/RangeInputType.cpp:
(WebCore::RangeInputType::handleMouseDownEvent):
* Source/WebCore/platform/graphics/controls/SwitchThumbPart.h:
* Source/WebCore/platform/graphics/controls/SwitchTrackPart.h:
* Source/WebCore/platform/graphics/mac/controls/SwitchThumbMac.mm:
(WebCore::SwitchThumbMac::draw):
* Source/WebCore/platform/graphics/mac/controls/SwitchTrackMac.mm:
(WebCore::SwitchTrackMac::draw):
* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::updateSwitchThumbPartForRenderer):
(WebCore::updateSwitchTrackPartForRenderer):
(WebCore::RenderTheme::isChecked const):
* Source/WebCore/rendering/RenderTheme.h:
(WebCore::RenderTheme::switchPointerTrackingMagnitudeProportion const):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: https://commits.webkit.org/271542@main
